cmd/openshift-install/gather.go

```diff
@@ -0,0 +1,231 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	assetstore "github.com/openshift/installer/pkg/asset/store"
+	"github.com/openshift/installer/pkg/terraform"
+	"github.com/openshift/installer/pkg/types"
+	awstypes "github.com/openshift/installer/pkg/types/aws"
+	libvirttypes "github.com/openshift/installer/pkg/types/libvirt"
+	openstacktypes "github.com/openshift/installer/pkg/types/openstack"
+)
+
+func newGatherCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "gather",
+		Short: "Gather debugging data for a given installation failure",
+		Long: `Gather debugging data for a given installation failure.
+
+When installation for Openshift cluster fails, gathering all the data useful for debugging can
+become a difficult task. This command helps users to collect the most relevant information that can be used
+to debug the installation failures`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmd.Help()
+		},
+	}
+	cmd.AddCommand(newGatherBootstrapCmd())
+	return cmd
+}
+
+var (
+	gatherBootstrapOpts struct {
+		bootstrap string
+		masters   []string
+	}
+)
+
+func newGatherBootstrapCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "bootstrap",
+		Short: "Gather debugging data for a failing to bootstrap control plane",
+		Args:  cobra.ExactArgs(0),
+		RunE:  runGatherBootstrapCmd,
+	}
+	cmd.PersistentFlags().StringVar(&gatherBootstrapOpts.bootstrap, "bootstrap", "", "")
+	cmd.PersistentFlags().StringArrayVar(&gatherBootstrapOpts.masters, "master", []string{}, "")
+	return cmd
+}
+
+func runGatherBootstrapCmd(_ *cobra.Command, _ []string) error {
+	cleanup := setupFileHook(rootOpts.dir)
+	defer cleanup()
+
+	tfStateFilePath := filepath.Join(rootOpts.dir, terraform.StateFileName)
```

Comment:
I'd much rather have the bootstrap IP stored somewhere more isolated from Terraform.  @abhinavdahiya had previously discussed storing it in `metadata.json` (although it's been a while; my memory is fuzzy).  But we could have `Cluster.Generate` add a per-platform switch so each platform could provide the bootstrap IP, and we could just drop that in a separate `created.json`.

Reply:
> @abhinavdahiya had previously discussed storing it in metadata.json (although it's been a while; my memory is fuzzy). But we could have Cluster.Generate add a per-platform switch so each platform could provide the bootstrap IP, and we could just drop that in a separate created.json, or some such.

do not want to expand the files created by installer anymore for now. the `tfstate` has everything we would need for now. This approach works fine for now as this allows us to get bootstrap ip and control plane ips for each platform.

Reply:
> This approach works fine for now as this allows us to get bootstrap ip and control plane ips for each platform.

But it requires all of our platforms to use Terraform?  Anyhow, maybe ok for the short term, but I'd at least like to push the per-platform code down into per-platform subpackages so we can delegate ownership.

Reply:
Another motivation for extracting this to Terraform (and maybe like @sferich888's question [here][1]), is "what about folks who want the bootstrap IP for other purposes?  For example, when we run the gather code from CI, it would be nice to be able to drop the Terraform -> bootstrap IP extraction and lean on this new installer code instead.

Reply:
> But it requires all of our platforms to use Terraform? Anyhow, maybe ok for the short term,

we have bigger problems as a lot of other places already assume terraform. so we solve that when we get to that bridge.

> but I'd at least like to push the per-platform code down into per-platform subpackages so we can delegate ownership.

personally this is so small and i don't see this growing that i'm fine keeping it here for now.

Reply:
> we have bigger problems as a lot of other places already assume terraform. so we solve that when we get to that bridge.

Works for me.

Reply:
What about failures where tfstate is not crated?

Reply:
> What about failures where tfstate is not crated?

This is gather on bootstrap failure so I'm not sure why you think tfstate will not exist in ipi

And on UPI where we know tfstate will never exist, we require user provide both bootstrap host and control plane host addresses